src/contracts/sgho/sGHO.sol

```diff
@@ -0,0 +1,380 @@
+// SPDX-License-Identifier: agpl-3
+pragma solidity ^0.8.19;
+
+import {Initializable} from 'openzeppelin-contracts-upgradeable/contracts/proxy/utils/Initializable.sol';
+import {Math} from 'openzeppelin-contracts/contracts/utils/math/Math.sol';
+import {SafeCast} from 'openzeppelin-contracts/contracts/utils/math/SafeCast.sol';
+import {ERC4626Upgradeable} from 'openzeppelin-contracts-upgradeable/contracts/token/ERC20/extensions/ERC4626Upgradeable.sol';
+import {ERC20PermitUpgradeable} from 'openzeppelin-contracts-upgradeable/contracts/token/ERC20/extensions/ERC20PermitUpgradeable.sol';
+import {ERC20Upgradeable} from 'openzeppelin-contracts-upgradeable/contracts/token/ERC20/ERC20Upgradeable.sol';
+import {AccessControlUpgradeable} from 'openzeppelin-contracts-upgradeable/contracts/access/AccessControlUpgradeable.sol';
+import {PausableUpgradeable} from 'openzeppelin-contracts-upgradeable/contracts/utils/PausableUpgradeable.sol';
+import {IERC20Permit} from 'openzeppelin-contracts/contracts/token/ERC20/extensions/IERC20Permit.sol';
+import {IERC20} from 'openzeppelin-contracts/contracts/token/ERC20/IERC20.sol';
+import {IERC4626} from 'openzeppelin-contracts/contracts/interfaces/IERC4626.sol';
+import {IERC20Metadata} from 'openzeppelin-contracts/contracts/token/ERC20/extensions/IERC20Metadata.sol';
+import {RescuableACL} from 'lib/aave-v3-origin/lib/solidity-utils/src/contracts/utils/RescuableACL.sol';
+import {RescuableBase, IRescuableBase} from 'lib/aave-v3-origin/lib/solidity-utils/src/contracts/utils/RescuableBase.sol';
+import {IsGHO} from './interfaces/IsGHO.sol';
+
+/**
+ * @title sGHO Token
+ * @author @kpk
+ * @notice sGHO is an ERC4626 vault that allows users to deposit GHO and earn yield.
+ * @dev This contract implements the ERC4626 standard for tokenized vaults, where the underlying asset is GHO.
+ * It also includes functionalities for yield generation based on a target rate, and administrative roles for managing the contract.
+ */
+contract sGHO is
+  Initializable,
+  ERC4626Upgradeable,
+  ERC20PermitUpgradeable,
+  AccessControlUpgradeable,
+  RescuableACL,
+  PausableUpgradeable,
+  IsGHO
+{
+  using Math for uint256;
+  using SafeCast for uint256;
+
+  // RAY is used for high-precision mathematical operations to avoid rounding errors
+  uint176 private constant RAY = 1e27;
+
+  /// @custom:storage-location erc7201:gho.storage.sGHO
+  struct sGHOStorage {
+    // Storage variables - Optimally packed for gas efficiency
+    uint176 yieldIndex; // 22 bytes - current yield index for share/asset conversion
+    uint64 lastUpdate; // 8 bytes - timestamp of last yield index update
+    uint16 targetRate; // 2 bytes - target annual yield rate in basis points (e.g., 1000 = 10%)
+    uint160 supplyCap; // 20 bytes - maximum total assets allowed in the vault
```

Comment:
is this in asset terms or whole units?

Reply:
because this is used elsewhere with total assets, i'm assuming it's in asset terms